BUG: Prevent Non-Manifold Geometry

DVigneault opened the pull request:
This patch addresses a long-standing issue in the Cuberille module
which frequently prevented its use with itk::QuadEdgeMesh, due a
number of cases in which this filter produces non-manifold geometry.
This patch determines, for each 2x2x2 neighborhood surrounding a
vertex, how many face-connected components N are present, and replicates
the relevant vertex N times.  Then, in the vertex map, the bitmask
and the position of the pixel index is used to determine which vertex
to select in the construction of each face.

DVigneault commented:
This patch begins to address [this issue](https://github.com/InsightSoftwareConsortium/ITKCuberille/issues).  The problem, briefly, is that when two foreground pixels abut by an edge or vertex, but not by a face, the `itk::Cuberille` class assigns the same vertex to each face, creating non-manifold geometry.  Here is a brief summary of the patch:

- It should be possible to avoid the above issue by "replicating" vertices when necessary (up to 4 times in the most pathological case of a "checkerboard" surrounding a vertex).
- If the 2x2x2 region surrounding a given vertex is extracted and given as an input to `itk::ConnectedComponentImageFilter` with `FullyConnected` set to `Off`, the number of connected components is equal to the number of times the vertex must be replicated, and quads derived from the same component should share a vertex.
- I may be costly to run this every time a new vertex was inserted.  However, there are only 256 possible combinations of foreground/background.  Therefore, it is possible to pre-compute the 256 possible binary masks, run connected components on them, and save the components array.  In this case, during execution, it is only necessary to match the local region to a given binary mask, and reference the correct component number.
- The component number is correlated to the index of the replicated point ID.

This works well so far in my local "torture tests."  However, the current test suite checks the number of points in the output mesh.  Therefore, whenever the above case occurs, it does a test failure:

```
	  3 - Cuberille_Blob2_01 (Failed)
	  4 - Cuberille_Blob3_01 (Failed)
	  6 - Cuberille_MarschnerLobb_01 (Failed)
	 11 - Cuberille_Neghip_01 (Failed)
	 12 - Cuberille_Neghip_02 (Failed)
	 13 - Cuberille_Neghip_03 (Failed)
```

An example test failure output is:

```
3: Error: Expected mesh with 14 points, but found 16
```

I hope to check these test cases soon, in order to determine whether this truly is the only source of the errors, or whether this patch introduces any additional problems.

In the meantime, any feedback is greatly appreciated.

DVigneault commented:
I've carefully inspected the output meshes (converted to obj) in Blender, and convinced myself that the output are correct.  I've updated the number of points asserted in the tests, and the tests are now passing.  Would love to get the community's opinion on this!

dzenanz commented on include/itkCuberilleImageToMeshFilter.hxx:
```diff
@@ -247,12 +255,13 @@ CuberilleImageToMeshFilter<TInputImage,TOutputMesh,TInterpolator>
              const TInputImage* image,
-             TOutputMesh* mesh )
+             TOutputMesh* mesh,
+             const size_t num_components )
```
ITK's style guide call for `camelCase`, and no `names_with_underscores`. It appears in a few more places in this PR.

dzenanz commented on include/itkCuberilleImageToMeshFilter.hxx:
```diff
@@ -535,6 +549,81 @@ CuberilleImageToMeshFilter<TInputImage,TOutputMesh,TInterpolator>
+template<typename TInputImage, typename TOutputMesh, typename TInterpolator>
+void
+CuberilleImageToMeshFilter<TInputImage,TOutputMesh,TInterpolator>
+::CalculateLabelsArray()
```
How long does it take to run this?

DVigneault commented:
@dzenanz Thanks for the review!  Fixed the snake case variables, and added timing to the test suite.  The method takes about 0.03s to run on my machine (Macbook Pro).  Another option would be to hard-code the values, which I'm also happy to do--let me know what you think.  As for timing, here is the before and after (again, on my Macbook).  

`master` branch:

```
      Start  1: Cuberille_Blob0_00
1: Polygonization took 0.00112295 seconds
      Start  2: Cuberille_Blob1_01
2: Polygonization took 0.000997066 seconds
      Start  3: Cuberille_Blob2_01
3: Polygonization took 0.00117993 seconds
      Start  4: Cuberille_Blob3_01
4: Polygonization took 0.00097394 seconds
      Start  5: Cuberille_Blob4_01
5: Polygonization took 0.0109069 seconds
      Start  6: Cuberille_MarschnerLobb_01
6: Polygonization took 0.066447 seconds
      Start  7: Cuberille_Fuel_01
7: Polygonization took 0.00786996 seconds
      Start  8: Cuberille_Fuel_02
8: Polygonization took 0.022701 seconds
      Start  9: Cuberille_Fuel_03
9: Polygonization took 0.0240922 seconds
      Start 10: Cuberille_HydrogenAtom_01
10: Polygonization took 0.103101 seconds
      Start 11: Cuberille_Neghip_01
11: Polygonization took 0.015518 seconds
      Start 12: Cuberille_Neghip_02
12: Polygonization took 0.0520141 seconds
      Start 13: Cuberille_Neghip_03
13: Polygonization took 0.0570581 seconds
      Start 14: Cuberille_Nucleon_01
14: Polygonization took 0.00340009 seconds
      Start 15: Cuberille_Nucleon_02
15: Polygonization took 0.0123041 seconds
      Start 16: Cuberille_Nucleon_03
16: Polygonization took 0.0128748 seconds
      Start 17: Cuberille_Silicium_01
17: Polygonization took 0.0151429 seconds
      Start 18: Cuberille_Silicium_02
18: Polygonization took 0.064738 seconds
      Start 19: Cuberille_Silicium_03
19: Polygonization took 0.0689571 seconds
      Start 20: CuberilleTestDirectionMatrix
      Start 21: CuberilleTestCellData
```

`PreventNonmanifoldGeometry` branch:

```
      Start  1: Cuberille_Blob0_00
1: Time to calculate labels array: 0.034729
1: Polygonization took 0.000927925 seconds
      Start  2: Cuberille_Blob1_01
2: Time to calculate labels array: 0.0345929
2: Polygonization took 0.000937939 seconds
      Start  3: Cuberille_Blob2_01
3: Time to calculate labels array: 0.034811
3: Polygonization took 0.000931978 seconds
      Start  4: Cuberille_Blob3_01
4: Time to calculate labels array: 0.0362709
4: Polygonization took 0.00108194 seconds
      Start  5: Cuberille_Blob4_01
5: Time to calculate labels array: 0.0349941
5: Polygonization took 0.01174 seconds
      Start  6: Cuberille_MarschnerLobb_01
6: Time to calculate labels array: 0.0350101
6: Polygonization took 0.0782831 seconds
      Start  7: Cuberille_Fuel_01
7: Time to calculate labels array: 0.0346479
7: Polygonization took 0.0117531 seconds
      Start  8: Cuberille_Fuel_02
8: Time to calculate labels array: 0.034817
8: Polygonization took 0.025409 seconds
      Start  9: Cuberille_Fuel_03
9: Time to calculate labels array: 0.035161
9: Polygonization took 0.0266399 seconds
      Start 10: Cuberille_HydrogenAtom_01
10: Time to calculate labels array: 0.0351291
10: Polygonization took 0.119708 seconds
      Start 11: Cuberille_Neghip_01
11: Time to calculate labels array: 0.0348642
11: Polygonization took 0.0243342 seconds
      Start 12: Cuberille_Neghip_02
12: Time to calculate labels array: 0.037019
12: Polygonization took 0.062902 seconds
      Start 13: Cuberille_Neghip_03
13: Time to calculate labels array: 0.0345449
13: Polygonization took 0.0665371 seconds
      Start 14: Cuberille_Nucleon_01
14: Time to calculate labels array: 0.0347452
14: Polygonization took 0.0056169 seconds
      Start 15: Cuberille_Nucleon_02
15: Time to calculate labels array: 0.0357108
15: Polygonization took 0.0137749 seconds
      Start 16: Cuberille_Nucleon_03
16: Time to calculate labels array: 0.0350361
16: Polygonization took 0.0151219 seconds
      Start 17: Cuberille_Silicium_01
17: Time to calculate labels array: 0.0348289
17: Polygonization took 0.029016 seconds
      Start 18: Cuberille_Silicium_02
18: Time to calculate labels array: 0.035764
18: Polygonization took 0.078846 seconds
      Start 19: Cuberille_Silicium_03
19: Time to calculate labels array: 0.0369978
19: Polygonization took 0.085942 seconds
      Start 20: CuberilleTestDirectionMatrix
      Start 21: CuberilleTestCellData
```

<img width="650" alt="image" src="https://user-images.githubusercontent.com/10100550/73879501-9abb1680-482a-11ea-89c0-5d51b75359e1.png">

dzenanz commented on include/itkCuberilleImageToMeshFilter.hxx:
```diff
@@ -535,6 +549,81 @@ CuberilleImageToMeshFilter<TInputImage,TOutputMesh,TInterpolator>
+
+    const auto connected = TConnected::New();
+    connected->SetInput( image );
+    connected->FullyConnectedOff();
```
The construction can probably be taken out of the loop, and there should be small performance benefit to it.

DVigneault commented:
Thanks @dzenanz!  There is still more I'd like to do in terms of performance and testing, but I think this might be enough for one PR.  Removing WIP.
